src/plot_api/plot_api.js

```diff
@@ -1561,13 +1614,13 @@ Plotly.restyle = function restyle(gd, astr, val, traces) {
             if(['orientation', 'type'].indexOf(ai) !== -1) {
                 axlist = [];
                 for(i = 0; i < traces.length; i++) {
-                    var trace = gd.data[traces[i]];
+                    var trace = data[traces[i]];
 
                     if(Registry.traceIs(trace, 'cartesian')) {
                         addToAxlist(trace.xaxis || 'x');
                         addToAxlist(trace.yaxis || 'y');
 
-                        if(astr === 'type') {
+                        if(ai === 'type') {
```

Comment:
That fixes an undiscoved bug.

`ai` is more general than `astr`

`astr` is the second argument of `Plotly.restyle(gd, astr, val)`, wherease `ai` is a key in `{ astr: val }` object constructed before this loop.